Fix observation session query

willmurphyscode opened the pull request:
This fix caches the `ObservationSession` instance where possible to prevent running a select query against observation sessions every time an observation is inserted. 

It also adds some more JSON observation objects to the test data so that redundant query work will be easier to see in test logs, if it ever comes up again.

willmurphyscode committed:
removes n+1 query. clean up of code coming soon

willmurphyscode committed:
refactor to eliminate long trainwreck

rposborne commented on app/controllers/api/observation_sessions_controller.rb:
```diff
@@ -2,16 +2,24 @@ module Api
             record.details = observation
-            record.observation_session_id = observation['observation_session_id']
+            record.observation_session = session
+            record.save
```
Is this save required?  Will find_or_create save after we leave the block passed to it?

willmurphyscode replied:
I believe find or create will execute the block if it's a create, but not if it's a find, but because the block is passed on to create, the value will be set anyway. I can definitely remove the `save`, but I'm concerned the method doesn't behave as expected when an observation with that ID is sent. https://stackoverflow.com/q/5461315/3896861

Does this method ever need to update an existing observation id?

rposborne replied:
Hmm, that's a good point.  I think since the data is immutable from the device that this is OKAY.  It's certainly something that we might want to note.   Does leaving the save their trigger a double save?

rposborne commented on app/controllers/api/observation_sessions_controller.rb:
```diff
@@ -2,16 +2,24 @@ module Api
-            record.observation_session_id = observation['observation_session_id']
+            record.observation_session = session
+            record.save
+            processed_observations << record
```
Could we just map the outer `Array#each` vs making it our selves?

willmurphyscode replied:
Yes. I'll make that change.

rposborne reviewed (changes requested):
Minor code cleanup,  It might be best to run this through rubocop as well there are a couple style tweaks for the project.

willmurphyscode committed:
Aesthetic Refactor

Remove the redundate `save` call from the block passed to
`find_or_create_by` and change `each` to `map` to avoid an unnecessary
array.

willmurphyscode commented:
I ran rubocop - the only thing it complains about in `observation_session_controller` is that the `create` method is too long, both by line count and Abc size. Is that something you'd like me to refactor out?

rposborne commented:
@willmurphyscode ABC / length does not need to be refactored out.
